File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        def expand(l, r):
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            return s[l+1:r]

        longest = ""
        for i in range(len(s)):
            temp1 = expand(i, i)
            temp2 = expand(i, i+1)

            if len(temp1) > len(longest):
                longest = temp1
            if len(temp2) > len(longest):
                longest = temp2

        return longest

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        if (n < 2) return s;

        int start = 0, maxLen = 1;

        auto expand = [&](int l, int r) {
            while (l >= 0 && r < n && s[l] == s[r]) {
                l--;
                r++;
            }
            return r - l - 1;
        };

        for (int i = 0; i < n; i++) {

            int len1 = expand(i, i);     // odd
            int len2 = expand(i, i+1);   // even

            int len = max(len1, len2);

            if (len > maxLen) {
                maxLen = len;
                start = i - (len-1)/2;
            }
        }

        return s.substr(start, maxLen);
    }
};
